### PR TITLE
Update S4043 (`no-misleading-array-reverse`): Report on `splice` and suggest `toSpliced`

### LINKED ISSUE #3933: Update S4043 (`no-misleading-array-reverse`): Report on `splice` and suggest `toSpliced`
We can extend the scope of the rule with the method [Array.prototype.splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) which works in place.

Furthermore, similar to https://github.com/SonarSource/SonarJS/issues/3865, we can suggest using the newly introduced [Array.prototype.toSpliced()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced) which returns a new array.

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3933

### CODE REVIEW

**Review by ilia-kebets-sonarsource — CHANGES_REQUESTED** [review]
See my proposal

File: src/linting/eslint/rules/no-misleading-array-reverse.ts [hunk 1]
```diff
@@ -62,7 +70,18 @@ export const rule: Rule.RuleModule = {
               isForbiddenOperation(node)
             ) {
               const method = formatMethod(propertyText);
-              const suggestedMethod = method === 'sort' ? 'toSorted' : 'toReversed';
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply src/linting/eslint/rules/no-misleading-array-reverse.ts:65]: I am not sure I want to open the Pandora's box of exceptions. If we add an exception for this particular use case, I am afraid we will start adding more. Furthermore, I think this snippet is a perfect example of unreadable code. This is why I would avoid doing anything here. Does it make sense?

[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply src/linting/eslint/rules/no-misleading-array-reverse.ts:65]: Agreed that the example is not the best code, but how would you do the following: remove an element from an array and perform some computation on it?

```javascript
const elem = array.splice(pos, 1);
process(elem)
```

Without the exception, we would be forced to do:

```javascript
const elem = array[pos]
array.splice(pos, 1);
process(elem)
```

The process would be like that when deleting >1 elements, which might indeed be nice to uniformize. I'll have a look on sourcegraph.

File: src/linting/eslint/rules/no-misleading-array-reverse.ts [hunk 2]
```diff
@@ -32,7 +32,15 @@ import {
   localAncestorsChain,
 } from './helpers';
 
-const arrayMutatingMethods = ['reverse', "'reverse'", '"reverse"', ...sortLike];
+const arrayMutatingMethods = [
+  'splice',
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply src/linting/eslint/rules/no-misleading-array-reverse.ts]: 👍

File: src/linting/eslint/rules/no-misleading-array-reverse.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply src/linting/eslint/rules/no-misleading-array-reverse.ts:65]: OK, nevermind, most of the time, when 1 item is removed, it isn't used: https://sourcegraph.com/search?q=context:global+lang:JavaScript+/splice%5C%28%5Cd%2C%5Cs%3F1/&patternType=standard&sm=1&groupBy=repo

**Review by ilia-kebets-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**